l1-contracts/src/core/messagebridge/Inbox.sol

```diff
@@ -0,0 +1,148 @@
+
+  /// @notice Given a message, computes an entry key for the Inbox
+  function computeMessageKey(L1ToL2Msg memory message) public pure returns (bytes32) {
+    return keccak256(
+      abi.encode(
+        message.sender,
+        message.recipient,
+        message.content,
+        message.secretHash,
+        message.deadline,
+        message.fee
+      )
+    );
+  }
+
+  /**
+   * @notice Inserts an entry into the Inbox
+   * @dev Will emit `MessageAdded` with data for easy access by the sequencer
+   * @dev msg.value - The fee provided to sequencer for including the entry
+   * @param _recipient - The recipient of the entry
+   * @param _deadline - The deadline to consume a message. Only after it, can a message be cancalled.
+   * @param _content - The content of the entry (application specific)
+   * @param _secretHash - The secret hash of the entry (make it possible to hide when a specific entry is consumed on L2)
+   * @return The key of the entry in the set
+   */
+  function sendL2Message(
+    L2Actor memory _recipient,
+    uint256 _deadline,
+    bytes32 _content,
+    bytes32 _secretHash
+  ) external payable returns (bytes32) {
+    L1ToL2Msg memory message = L1ToL2Msg({
+      sender: L1Actor(msg.sender, block.chainid),
+      recipient: _recipient,
+      content: _content,
+      secretHash: _secretHash,
+      deadline: _deadline,
+      fee: msg.value
+    });
+
+    bytes32 key = computeMessageKey(message);
+    _insert(key);
+
+    emit MessageAdded(
+      key,
+      message.sender.actor,
+      message.recipient.actor,
+      message.sender.chainId,
+      message.recipient.version,
+      message.deadline,
+      message.fee,
+      message.content
+    );
+
+    return key;
+  }
+
+  /**
+   * @notice Cancel a pending L2 message
+   * @dev Will revert if the deadline have not been crossed - message only cancellable past the deadline
+   *  so it cannot be yanked away while the sequencer is building a block including it
+   * @dev Must be called by portal that inserted the entry
+   * @param _message - The content of the entry (application specific)
+   * @param _feeCollector - The address to receive the "fee"
+   * @return The key of the entry removed
+   */
+  function cancelL2Message(L1ToL2Msg memory _message, address _feeCollector)
+    external
+    returns (bytes32)
+  {
+    if (msg.sender != _message.sender.actor) revert Inbox__Unauthorized();
+    if (_message.deadline <= block.timestamp) revert Inbox__NotPastDeadline();
+    return _consumeInboxMessage(_message, _feeCollector);
+  }
+
+  /**
+   * @notice Consumes an entry from the Inbox
+   * @dev Only callable by the rollup contract
+   * @dev Will revert if the message is already past deadline
+   * @param _message - The content of the entry (application specific)
+   * @param _feeCollector - The address to receive the "fee"
+   * @return The key of the entry removed
+   */
+  function consume(L1ToL2Msg memory _message, address _feeCollector)
```

Comment:
Rollup will not know all the stuff in `L1ToL2Msg`, it will just have the entry-key, otherwise costs explodes. Only need the "opening" for cancelling or consuming by the recipient (e.g., on L2 side). 

Because the rollup will execute it for a full block, consider using a list of entry keys instead.

Reply:
To handle your extra fields for deadlines, you are also running into issues with needing to pass more information to check the deadlines you want. Data which is not part of the blocks.